wpilibj/src/main/java/edu/wpi/first/wpilibj/SharpIR.java

```diff
@@ -0,0 +1,146 @@
+   */
+  public static SharpIR GP2Y0A02YK0F(int channel) {
+    return new SharpIR(channel, 62.28, -1.092, 20, 150.0);
+  }
+
+  /**
+   * Sharp GP2Y0A21YK0F is an analog IR sensor capable of measuring distances from 10cm to 80cm.
+   *
+   * @param channel Analog input channel the sensor is connected to
+   * @return sensor object
+   */
+  public static SharpIR GP2Y0A21YK0F(int channel) {
+    return new SharpIR(channel, 26.449, -1.226, 10.0, 80.0);
+  }
+
+  /**
+   * Sharp GP2Y0A41SK0F is an analog IR sensor capable of measuring distances from 4cm to 40cm.
+   *
+   * @param channel Analog input channel the sensor is connected to
+   * @return sensor object
+   */
+  public static SharpIR GP2Y0A41SK0F(int channel) {
+    return new SharpIR(channel, 12.84, -0.9824, 4.0, 40.0);
+  }
+
+  /**
+   * Manually construct a SharpIR object. The distance is computed using this formula: A*v ^ B.
+   * Prefer to use one of the static factories to create this device instead.
+   *
+   * @param channel AnalogInput channel
+   * @param a Constant A
+   * @param b Constant B
+   * @param minCM Minimum distance to report in centimeters
+   * @param maxCM Maximum distance to report in centimeters
+   */
+  public SharpIR(int channel, double a, double b, double minCM, double maxCM) {
+    m_sensor = new AnalogInput(channel);
+
+    m_A = a;
+    m_B = b;
+    m_minCM = minCM;
+    m_maxCM = maxCM;
+
+    SendableRegistry.addLW(this, "SharpIR", channel);
+
+    m_simDevice = SimDevice.create("SharpIR", m_sensor.getChannel());
+    if (m_simDevice != null) {
+      m_simRange = m_simDevice.createDouble("Range (cm)", Direction.kInput, 0.0);
+      m_sensor.setSimDevice(m_simDevice);
+    }
+  }
+
+  @Override
+  public void close() {
+    SendableRegistry.remove(this);
+    m_sensor.close();
+    m_sensor = null;
+
+    if (m_simDevice != null) {
+      m_simDevice.close();
+      m_simDevice = null;
+      m_simRange = null;
+    }
+  }
+
+  /**
+   * Get the analog input channel number.
+   *
+   * @return analog input channel
+   */
+  public int getChannel() {
+    return m_sensor.getChannel();
+  }
+
+  /**
+   * Get the range in inches from the distance sensor.
+   *
+   * @return range in inches of the target returned by the sensor
+   */
+  public double getRangeInches() {
+    return getRangeCM() / 2.54;
+  }
+
+  /**
+   * Get the range in centimeters from the distance sensor.
+   *
+   * @return range in centimeters of the target returned by the sensor
+   */
+  public double getRangeCM() {
+    double distance;
+
+    if (m_simRange != null) {
+      distance = m_simRange.get();
+    } else {
+      // Don't allow zero/negative values
+      var v = Math.max(m_sensor.getVoltage(), 0.00001);
+      distance = m_A * Math.pow(v, m_B);
+    }
+
+    // Always constrain output
+    return Math.max(Math.min(distance, m_maxCM), m_minCM);
+  }
+
+  @Override
+  public void initSendable(SendableBuilder builder) {
+    builder.setSmartDashboardType("SharpIR");
```

Comment:
In my team's ShariIR class, we used Ultrasonic so that new dashboard widgets didn't have to be created

Reply:
Does it actually do anything useful?